### DISCUSSION TITLE
Use with Other Control Vendors?

### DISCUSSION THREAD

**Discussion by benhysell** [CommunityToolkit/Maui.Markup#86] [Q&A] [answered]
Can Maui.Markup be used with other control vendor's controls, i.e. Telerik, DevExpress, Syncfusion?

I started playing with Telerik and did not have a lot of luck, I hope I'm missing something simple.

```
var stackLayoutStopStartButtons = new StackLayout();
var dataGridHashedData = new RadDataGrid();
stackLayoutStopStartButtons.Add(dataGridHashedData);
```
The last line - Error CS1503	Argument 1: cannot convert from 'Telerik.UI.Xaml.Controls.Grid.RadDataGrid' to 'Microsoft.Maui.IView'	V.RelayDatabase (net6.0-windows10.0.19041.0)	

Thanks!

**Comment by TheCodeTraveler** [comment, marked as answer]
> Can Maui.Markup be used with other control vendor's controls, i.e. Telerik, DevExpress, Syncfusion?

Short answer: Yes. 

However, the bug you're getting has nothing to do with `CommunityToolkit.Maui.Markup`; there's no code referencing the `CommunityToolkit.Maui.Markup` library in your sample. 

The problem is that `Telerik`'s implementation of `RadDataGrid` didn't implement `IView`. I recommend reaching out to `Telerik` for assistance with their library.

**Reply by LanceMcCarthy** [reply]
Hi Brandon! 

For the controls we still have as Contract and not a Handler, we use `Microsoft.Maui.Controls.Layout`, which is implementing `IView`.

![image](https://user-images.githubusercontent.com/3520532/179583804-27282f77-cda3-4048-a458-58db426bf2c3.png)

Are you saying that every type must directly implement IView and Maui.Layout is not a valid base for fluent builder expressions? Or are you saying that control through Contracts are not supported at all (i.e., they all must be 100% Handlers)?

@benhysell The control type in the error you're seeing are actually the final native control handled by the compatibility Contract. I will follow up with you in our support conversation about the difference between a contract and a handler, as well as our road map for re-writing every control as a Handler.

**Reply by bijington** [reply]
@LanceMcCarthy I believe Brandon's point referred to the fact that the code in the original sample is .NET MAUI code (calling `StackLayout.Add` here https://github.com/dotnet/maui/blob/main/src/Controls/src/Core/Layout/Layout.cs) rather than any code on this toolkit.

Your inheritance hierarchy looks like the code should work to me in that scenario. I haven't tried using the Telerik package though. I can see if I can reproduce it

**Reply by LanceMcCarthy** [reply]
Hi @bijington,  TBH I haven't attempted it myself yet, was just doing some initial triage. Now that it seems there might be some cause to break out a test project and prepare something. Let me handle the test setup for you and pass off the project once I have something you can use (~30 minutes).

**Reply by LanceMcCarthy** [reply]
@benhysell & @bijington This works fine for me:

![image](https://user-images.githubusercontent.com/3520532/179608914-38c2a989-f83b-4740-9d03-c41a7d4f1dc4.png)

```csharp
using Telerik.Maui.Controls.Compatibility.DataGrid;

namespace FluentBuilderTest;

public class NewPage1 : ContentPage
{
	public NewPage1()
    {
        var dg = new RadDataGrid
        {
            ItemsSource = Enumerable.Range(1, 100).Select(i => new Car
            {
                Make = $"Make {i}",
                Model = $"Model {i}",
                Year = 1920 + i
            })
        };

        // Test 1 - Grid child
        Grid grid = new Grid();
        grid.Children.Add(dg);
        
        Content = grid;
    }
}

public class Car
{
    public string Make { get; set; }
    public string Model { get; set; }
    public int Year { get; set; }
}
```

Ben, make sure you've followed our getting started docs and the project has registered the MAUI handlers. Did you skip step 4 here? https://docs.telerik.com/devtools/maui/get-started/first-steps#step-4-register-required-handlers-and-renderers. I will reply to the Telerik Feature Request you opened with this test project so you can compare.

**Reply by bijington** [reply]
That is good to hear as the code looked like it should behave. To confirm none of the above is using this Markup Toolkit, if it were it would look a bit more like:

```csharp
public NewPage1()
{       
    Content = new Grid
    {
        Children =
        {
            new RadDataGrid
            {
                ItemsSource = Enumerable.Range(1, 100).Select(i => new Car
                {
                    Make = $"Make {i}",
                    Model = $"Model {i}",
                    Year = 1920 + i
                })
            }
        }
    };
}
```

Even the above doesn't actually use our toolkit, if we started to add more rows or columns then you can fluently set those as per the docs at: https://docs.microsoft.com/en-gb/dotnet/communitytoolkit/maui/markup/extensions/grid-extensions

Of course there may eventually be scenarios where developers will want to fluently set properties on Telerik or other vendor controls but if they don't make use of the MAUI interfaces then I am not convinced we would support it as part of this toolkit.

**Reply by LanceMcCarthy** [reply]
Agreed, I also mentioned that in my final response here https://feedback.telerik.com/maui/1572850-work-with-https-github-com-communitytoolkit-maui-markup. I don't know why this Issue was opened in the Maui.Markup repo. 

I think it's just a case of misunderstanding the error and not knowing where to go next for support. Pointing at the relationship between the toolkit and the controls as a potential conflict source seems like an initial false flag.

Regardless, I've requested a repro project demonstrating any further issues just in case something arises while implementing any of the Maui.Markup extension methods. In which case, we (Progress) can address independently on a property by property basis in Telerik UI for Maui.

Edit:
I did try some quick extensions to eliminate any kind of factor that triggers an exception from tryign to use the interface directly and it seems fine, too:

![image](https://user-images.githubusercontent.com/3520532/179614771-3113265d-e695-4857-ac57-53a730fc97a3.png)

**Comment by TheCodeTraveler** [comment]
Converting this Issue to a Discussion as it is not a bug in `Maui.Markup`.

**Comment by CliffAgius** [comment]
Closed and answered
